pyquil/quilatom.py

```diff
@@ -899,26 +900,58 @@ def out(self) -> str:
         return str(self)
 
 
-@dataclass
-class WaveformReference(QuilAtom):
+class WaveformReference(quil_rs.WaveformInvocation):
     """
     Representation of a Waveform reference.
     """
 
-    name: str
-    """ The name of the waveform. """
+    def __new__(cls, name: str):
+        return super().__new__(cls, name, {})
 
     def out(self) -> str:
-        return self.name
+        return str(self)
+
+
+def _template_waveform_property(name: str, doc: Optional[str] = None):
+    """
+    Helper method for initializing getters, setters, and deleters for
+    parameters on a ``TemplateWaveform``. Should only be used inside of
+    ``TemplateWaveform`` or one its base classes.
+    """
+
+    def fget(self: "TemplateWaveform"):
+        return self.get_parameter(name)
+
+    def fset(self: "TemplateWaveform", value: ParameterDesignator):
+        self.set_parameter(name, value)
+
+    def fdel(self: "TemplateWaveform"):
+        self.set_parameter(name, None)
+
+    return property(fget, fset, fdel, doc)
```

Comment:
Would've liked this to exist as a `staticmethod` on `TemplateWaveform` but it's not possible on Python <= 3.9 without doing some more hacky nonsense.

Reply:
another piece added to the python lightsaber you're building